File: libs/hikyuu_python/trade_sys/_Stoploss.cpp
/*
 * _Stoploss.cpp
 *
 *  Created on: 2013-3-21
 *      Author: fasiondog
 */

#include <boost/python.hpp>
#include <hikyuu/trade_sys/stoploss/build_in.h>
#include "../_Parameter.h"
#include "../pickle_support.h"

using namespace boost::python;
using namespace hku;

class StoplossWrap : public StoplossBase, public wrapper<StoplossBase> {
public:
    StoplossWrap(): StoplossBase() {}
    StoplossWrap(const string& name): StoplossBase(name) {}
    virtual ~StoplossWrap() {}

    void _reset() {
        if (override func = this->get_override("_reset")) {
            func();
        } else {
            StoplossBase::_reset();
        }
    }

    void default_reset() {
        this->StoplossBase::_reset();
    }

    StoplossPtr _clone() {
        return this->get_override("_clone")();
    }

    void _calculate() {
        this->get_override("_calculate")();
    }

    price_t getPrice(const Datetime& datetime, price_t price) {
        return this->get_override("getPrice")(datetime, price);
    }

    price_t getShortPrice(const Datetime& datetime, price_t price) {
        if (override getShortPrice = get_override("getShortPrice")) {
            return getShortPrice(datetime, price);
        }
        return StoplossBase::getShortPrice(datetime, price);
    }

    price_t default_getShortPrice(const Datetime& datetime, price_t price) {
        return this->StoplossBase::getShortPrice(datetime, price);
    }
};

string (StoplossBase::*st_get_name)() const = &StoplossBase::name;
void (StoplossBase::*st_set_name)(const string&) = &StoplossBase::name;

void export_Stoploss() {
    class_<StoplossWrap, boost::noncopyable>("StoplossBase", init<>())
            .def(init<const string&>())
            .def(self_ns::str(self))
            .add_property("name", st_get_name, st_set_name)
            .def("getParam", &StoplossBase::getParam<boost::any>)
            .def("setParam", &StoplossBase::setParam<object>)
            .def("setTM", &StoplossBase::setTM)
            .def("getTM", &StoplossBase::getTM)
            .def("setTO", &StoplossBase::setTO)
            .def("getTO", &StoplossBase::getTO)
            .def("getPrice", pure_virtual(&StoplossBase::getPrice))
            .def("getShortPrice", &StoplossBase::getShortPrice,
                    &StoplossWrap::default_getShortPrice)
            .def("reset", &StoplossBase::reset)
            .def("clone", &StoplossBase::clone)
            .def("_calculate", pure_virtual(&StoplossBase::_calculate))
            .def("_reset", &StoplossBase::_reset, &StoplossWrap::default_reset)
            .def("_clone", pure_virtual(&StoplossBase::_clone))
#if HKU_PYTHON_SUPPORT_PICKLE
            .def_pickle(name_init_pickle_suite<StoplossBase>())
#endif
            ;

    register_ptr_to_python<StoplossPtr>();

    def("ST_FixedPercent", ST_FixedPercent, (arg("p") = 0.03));
    def("ST_Indicator", ST_Indicator, (arg("op"), arg("kpart") = "CLOSE"));
    def("ST_Saftyloss", ST_Saftyloss,
            (arg("n1") = 10, arg("n2") = 3, arg("p") = 2.0));
}

File: tools/hikyuu/test/Stoploss.py
#!/usr/bin/python
# -*- coding: utf8 -*-
# gb18030

#===============================================================================
# 作者：fasiondog
# 历史：1）20130321, Added by fasiondog
#===============================================================================

import unittest

from test_init import *
from hikyuu.trade_sys.stoploss import *

class StoplossPython(StoplossBase):
    def __init__(self):
        super(StoplossPython, self).__init__()
        self._x = 0
        
    def name(self):
        return "StoplossPython"
    
    def getPrice(self, datetime, price):
        if self._x < 10:
            return 0.0
        return 1.0
    
    def _reset(self):
        self._x = 0
            
    def _clone(self):
        p = StoplossPython()
        p._x = self._x
        return p
    
    def _calculate(self):
        """ do nothin """

class StoplossTest(unittest.TestCase):
    def test_StoplossBase(self):
        p = StoplossPython()
        self.assertEqual(p.name(), "StoplossPython")
        self.assertEqual(p.getPrice(Datetime(200101010000), 1.0), 0.0)
        
        self.assertEqual(p._x, 0)
        p._x = 10
        self.assertEqual(p._x, 10)
        self.assertEqual(p.getPrice(Datetime(200101010000), 1.0), 1.0)
        p.reset()
        self.assertEqual(p._x, 0)
        
        p._x = 10
        p_clone = p.clone()
        self.assertEqual(p_clone._x, 10)
        p.reset()
        self.assertEqual(p._x, 0)
        self.assertEqual(p_clone._x, 10)

                 
def suite():
    return unittest.TestLoader().loadTestsFromTestCase(StoplossTest)